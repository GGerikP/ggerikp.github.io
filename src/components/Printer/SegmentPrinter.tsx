import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import Link from '../Link';
import Cursor, { CursorDisplay } from './Cursor';

const SegmentWrapper = styled.span`
`;

const StyledLink = styled(Link)`
    text-decoration: underline;
`;

const PrintedTextSpan = styled.span`
`;

export type Segment = {
    text: string;
    link?: string;
    prePrintDelay?: number;
    postPrintDelay?: number;
}

type SegmentPrinterProps = {
    segment: Segment;
    segmentIndex: number;
    isLastSegment: boolean;
    isLastLine: boolean;
    typingSpeed: number;
    instantPrint: boolean;
    printNextSegment: () => void;
};

function SegmentPrinter ({ segment, segmentIndex, isLastSegment, isLastLine, typingSpeed, instantPrint, printNextSegment }: SegmentPrinterProps) {

  const [isDonePrinting, setIsDonePrinting] = useState<boolean>(false);
  const [charIndex, setCharIndex] = useState<number>(0);
  const [displayedText, setDisplayedText] = useState<string>('');
  const [cursorDisplay, setCursorDisplay] = useState<CursorDisplay>(CursorDisplay.Blink);
  const postPrintDelay = (!instantPrint && (segment.postPrintDelay) ? segment.postPrintDelay : 0);
  const prePrintDelay = !instantPrint ? (typingSpeed + ((charIndex === 0 && segment.prePrintDelay) ? segment.prePrintDelay : 0)) : 0;

  // Cursor Display
  useEffect(() => {
    // console.log(`isDonePrinting=${isDonePrinting}, isLastSegment=${isLastSegment}, isLastLine=${isLastLine}, cursorDisplay=${cursorDisplay}`);
    if (!isDonePrinting && (charIndex < segment?.text?.length)) {
      if (cursorDisplay !== CursorDisplay.Stable) {
        setCursorDisplay(CursorDisplay.Stable);
      }
    } else if (isDonePrinting && !(isLastSegment && isLastLine)) {
      if (cursorDisplay !== CursorDisplay.Hidden) {
        setCursorDisplay(CursorDisplay.Hidden);
      }
    } else if (cursorDisplay !== CursorDisplay.Blink) {
      setCursorDisplay(CursorDisplay.Blink);
    }
  }, [charIndex, isLastLine, isLastSegment, isDonePrinting, segment?.text?.length, cursorDisplay]);

  // Get Next Segment
  useEffect(() => {
    if (isDonePrinting) {
      printNextSegment();
    }
  }, [isDonePrinting, printNextSegment]);

  // Set whether we're done printing or not
  useEffect(() => {
    const postPrintTimeout = setTimeout(() => {
      if (charIndex >= segment?.text?.length) {
        setIsDonePrinting(true);
      }
    }, postPrintDelay);
    return (() => clearTimeout(postPrintTimeout));
  }, [segment, charIndex, postPrintDelay]);

  // Update the Displayed Text
  useEffect(() => {
    setDisplayedText(segment.text.substring(0, charIndex));
  }, [charIndex, segment]);

  // Print the next char
  useEffect(() => {
    const prePrintTimeout = setTimeout(() => {
      if (charIndex < segment?.text?.length) {
        setCharIndex(prevCharIndex => prevCharIndex + 1);
      }
    }, prePrintDelay);
    return (() => clearTimeout(prePrintTimeout));
  }, [segment, charIndex, prePrintDelay]);

  return (
    <SegmentWrapper id={`SegmentPrinter:${segmentIndex}`}>
      {segment.link
        ? <StyledLink
          url={segment.link}>
          <PrintedTextSpan>{displayedText}</PrintedTextSpan>
        </StyledLink>
        : <PrintedTextSpan>{displayedText}</PrintedTextSpan>
      }
      <Cursor cursorDisplay={cursorDisplay} index={segmentIndex} />
    </SegmentWrapper>
  );
}

export default SegmentPrinter;