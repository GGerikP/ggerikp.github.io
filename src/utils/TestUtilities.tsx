// Util Functions
export const checkIfCursorIsBlinking = (cursorElement: Element | null) => {
  const style = cursorElement ? getComputedStyle(cursorElement) : new CSSStyleDeclaration();
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: inline-block;');
  expect(cursorElement).toHaveStyle('border-left: 2px solid white;');
  expect(style.getPropertyValue('animation')).toMatch(/.*\.5s step-end 0s infinite.*/);
};

export const checkIfCursorIsHidden = (cursorElement: Element | null) => {
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: none;');
};

