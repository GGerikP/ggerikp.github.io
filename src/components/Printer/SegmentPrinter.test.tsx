import React from 'react';
import '@testing-library/jest-dom';
import SegmentPrinter, { Segment } from './SegmentPrinter';
import { render, waitFor } from '@testing-library/react';
import { checkIfCursorIsBlinking, checkIfCursorIsHidden } from '../../utils/TestUtilities';

describe('SegmentPrinter Tests', () => {

  const mockPrintNextSegment = jest.fn();

  it('Renders a single empty segment with no lines or segments to follow', async () => {
    const segment: Segment = {
      text: ''
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={true}
      isLastLine={true}
      typingSpeed={0}
      instantPrint={false}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 500 });

    // Check the cursor - should be blinking
    checkIfCursorIsBlinking(renderedComponent.container.querySelector('#Cursor\\:0'));
  }, 50000);

  it('Renders a single empty segment with more segments to follow', async () => {
    const segment: Segment = {
      text: ''
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={false}
      isLastLine={true}
      typingSpeed={0}
      instantPrint={false}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 500 });

    // Check the cursor - should be hidden
    checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));
  }, 50000);

  it('Renders a SegmentPrinter with a single character of text and it is the last segment and last line', async () => {
    const segment: Segment = {
      text: 'a'
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={true}
      isLastLine={true}
      typingSpeed={0}
      instantPrint={false}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(async () => {
      const element = await renderedComponent.findByText(/a.*/);
      expect(element).toBeInTheDocument();
    }, { timeout: 250 });
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 100 });

    // Check the cursor - should e blinking
    checkIfCursorIsBlinking(renderedComponent.container.querySelector('#Cursor\\:0'));
  });

  it('Renders a SegmentPrinter with a single character of text but there are more segments to print', async () => {
    const longText = 'AVeryLongStringPrintedOverAndOverAgain.';
    const segment: Segment = {
      text: longText
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={false}
      isLastLine={false}
      typingSpeed={0}
      instantPrint={false}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(async () => {
      const element = await renderedComponent.findByText(new RegExp(longText + '.*'));
      expect(element).toBeInTheDocument();
    }, { timeout: 500 });
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 250 });

    // Check the cursor
    checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));
  });

  it('Renders a SegmentPrinter with a single character of text but there are more segments to print', async () => {
    const longText = 'AVeryLongStringPrintedOverAndOverAgain.';
    const segment: Segment = {
      text: longText
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={false}
      isLastLine={true}
      typingSpeed={0}
      instantPrint={false}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(async () => {
      const element = await renderedComponent.findByText(new RegExp(longText + '.*'));
      expect(element).toBeInTheDocument();
    }, { timeout: 500 });
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 250 });

    // Check the cursor
    checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));
  });

  it('Renders a SegmentPrinter with a large string of text instantly with instantPrint set to true', async () => {

    const longText = 'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.' +
            'AVeryLongStringPrintedOverAndOverAgain.';
    const segment: Segment = {
      text: longText
    };
    const renderedComponent = render(<SegmentPrinter
      segment={segment}
      segmentIndex={0}
      isLastSegment={true}
      typingSpeed={0}
      instantPrint={true}
      isLastLine={true}
      printNextSegment={mockPrintNextSegment} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
    await waitFor(async () => {
      const element = await renderedComponent.findByText(new RegExp(longText + '.*'));
      expect(element).toBeInTheDocument();
    }, { timeout: 5000 });
    await waitFor(() => { expect(mockPrintNextSegment).toHaveBeenCalledTimes(1); }, { timeout: 100 });
  }, 10000);
});