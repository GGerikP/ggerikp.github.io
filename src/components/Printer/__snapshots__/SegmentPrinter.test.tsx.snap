// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentPrinter Tests Renders a SegmentPrinter with a large string of text instantly with instantPrint set to true 1`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU cFrJyp"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;

exports[`SegmentPrinter Tests Renders a SegmentPrinter with a single character of text and it is the last segment and last line 1`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU cFrJyp"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;

exports[`SegmentPrinter Tests Renders a SegmentPrinter with a single character of text but there are more segments to print 1`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU cFrJyp"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;

exports[`SegmentPrinter Tests Renders a SegmentPrinter with a single character of text but there are more segments to print 2`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU cFrJyp"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;

exports[`SegmentPrinter Tests Renders a single empty segment with more segments to follow 1`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU blcbix"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;

exports[`SegmentPrinter Tests Renders a single empty segment with no lines or segments to follow 1`] = `
<DocumentFragment>
  <span
    id="SegmentPrinter:0"
  >
    <span
      class="sc-dlfmHC gfsIAT"
    />
    <div
      class="sc-bdfCDU blcbix"
      id="Cursor:0"
    >
       
    </div>
  </span>
</DocumentFragment>
`;
